project_milestone/tests/test_project_milestone.py

```diff
@@ -0,0 +1,44 @@
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo.tests import common
+
+
+class TestProjectMilestone(common.TransactionCase):
+
+    def setUp(self):
+        super().setUp()
+        self.test_project = self.env['project.project'].create({
+            'name': 'TestProject'})
+        self.test_project_milestone_1 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_1',
+            'project_id': self.test_project.id})
+        self.test_project_milestone_2 = self.env['project.milestone'].create({
+            'name': 'TestMilestone_2',
+            'project_id': self.test_project.id})
+        self.test_open_stage = self.env['project.task.type'].create({
+            'name': 'TestOpenStage'})
+        self.test_close_stage = self.env['project.task.type'].create({
+            'name': 'TestCloseStage',
+            'closed': True})
+        self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_open_stage.id})
+        self.env['project.task'].create({
+            'name': 'TestTask',
+            'project_id': self.test_project.id,
+            'milestone_id': self.test_project_milestone_1.id,
+            'stage_id': self.test_close_stage.id})
+
+    def test_milestone_sequences(self):
+        milestone1 = self.test_project_milestone_1
+        milestone2 = self.test_project_milestone_2
+
+        self.assertEqual(milestone1.sequence, 1)
+        self.assertEqual(milestone2.sequence, 2)

```

Comment:
Sequences are by default implemented with the `CREATE SEQUENCE` feature from Postgres. Quoting from [their docs](https://www.postgresql.org/docs/current/sql-createsequence.html):

> `nextval` and `setval` calls are never rolled back

Your test suite now contains 2 tests, which are run like this:

1. Run 1
   - setUp -> milestones get sequences 1 and 2
   - test_milestone_progress -> 1st test, [ordered alphabetically](https://docs.python.org/3/library/unittest.html#organizing-test-code)
1. Run 2
   - setUp -> milestones get sequences 3 and 4
   - test_milestone_sequences -> 2nd test ordered alphabetically

Since the test order is not what you expect and the call to nextval is not rolled back, in the 2nd run, when `test_milestone_sequences` is run, they are created with sequences 3 and 4.

But the fix is not changing the method name, because not having into account test order is a good thing (these are unit tests after all). The fix also is not to check that sequences are 3 and 4, because a new unit test here with a lower name (alphabetically) could break it again easily.

The fix is:

```diff
-         self.assertEqual(milestone1.sequence, 1)
-         self.assertEqual(milestone2.sequence, 2)
+         self.assertGreater(milestone2.sequence, milestone1.sequence)
```

Because what actually matters to you is that the sequence of milestone 2 is greater than the sequence of milestone 1, regardless of their values.

Reply:
@Yajo you explain things very well, this makes sense to me. I think my error in my thinking was that the setup happens once and the other tests run using that single setup but it makes sense how you explained it. Thank you again, you've been very helpful.

Last is the group method that i'm still struggling to hit with the test. I'm just stumped on how to test that, even with reading @max3903 links. I guess i'm missing the concept maybe. I almost feel like I should remove it as maybe it's not really a requirement to show all the milestones on the group_by kanban. Thoughts?

Reply:
I am fine with merging as it is now. The test for grouping can come later.

Reply:
I agree with @max3903, although I think that getting 100% coverage will be a good practice for @patrickrwilson. He's learning this stuff after all.



> I almost feel like I should remove it as maybe it's not really a requirement to show all the milestones on the group_by kanban. Thoughts?

No, it's good as it is. The feature makes all milestones visible when grouping inside a project view; the same happens with i.e. stages, and it's quite helpful.

I'm adding a suggestion with the test code you possibly need.